### PR TITLE
add special case on isl_bool for callback return type's cast

### PR DESCRIPTION
**Pull request by thisiscam**
This implements a fix for isl_bool in callback handling (as mentioned in #30).

### CODE REVIEW

**Comment by inducer** [comment]
Thanks for working on this.

In #30, you suggested that

> Maybe the callback here also needs to catch exceptions? And if caught, return with `isl_bool_error`? 

I think that's a good idea, and it IMO works to the "spirit" of the interface. (But do make sure to still print the traceback so as to retain debuggability.)

> But then, I'm not really sure if this is the only place that needs a fix.

It's a step in the right direction, and there's no obviously "more general" place.

File: test/test_isl.py [hunk 1]
```diff
@@ -391,6 +391,20 @@ def test_ast_node_list_free():
...
+    umap1 = umap.remove_map_if(lambda m: False)
+    assert umap1 == umap, "map should not change"
+
+    umap2 = umap.remove_map_if(lambda m:
+        m.get_tuple_name(isl.dim_type.in_) == "B")
+    assert umap2 == isl.UnionMap("{A[0] -> [1]}")
+
+    def callback_throws_exception(m):
+        assert False
+    umap3 = umap.remove_map_if(callback_throws_exception)
```
Comment (thisiscam) [inline test/test_isl.py:406]: @inducer I don't know how to let the test infrastructure to "expect an exception"?

**Comment by thisiscam** [comment]
> (But do make sure to still print the traceback so as to retain debuggability.)

I think that this exception handler should be doing the error catching and printing. However the `Py_Error()` call is not working as expected (not printing anything). I suspect that maybe it is also not working for other callbacks.
https://github.com/inducer/islpy/blob/13e3c5695a4332cc401fa346f2ccb8b316bc3706/gen_wrap.py#L816

If so, this probably needs to go into another PR. 
The current PR performs correct casting, catches exceptions (with existing code on #L815), and is able to return to Python. However I also noticed that `m.remove_map_if()` corrupts `m` (if callback throws an exception) somehow and will result in segfault. Example:
```
import islpy
m = islpy.UnionMap("{[0] -> [1]}")
def raise_error(m):
  raise ValueError("xxx")
m1 = m.remove_map_if(raise_error)

# After above code:
# m1 is not defined, that's good
# access to m will cause Segmentation fault
print(m) 
```

File: test/test_isl.py (hunk 1 above)
Comment (kaushikcfd) [reply test/test_isl.py:406]: https://docs.pytest.org/en/reorganize-docs/new-docs/user/pytest_raises.html

**Comment by thisiscam** [comment]
I'm out of ideas of how to fix the issues above, and I'm not really familiar with how gen_wrap is dealing with exceptions. I think it's time for me to hand over control of this PR to the maintainer? @inducer @kaushikcfd. 

However, please let me know if I can be of more help!

**Comment by inducer** [comment]
> However the `PyErr_Print()` call is not working as expected (not printing anything).

7f856cd38c689a2b1b999bb0cedbd44e683a813e

**Comment by inducer** [comment]
For your second issue, I'm increasingly convinced that this is an isl bug. Using this diff:
```diff
-Subproject commit 13b9178b9b370164d0dbea232dbbe0112a27b6ae
+Subproject commit 13b9178b9b370164d0dbea232dbbe0112a27b6ae-dirty
diff --git a/src/wrapper/wrap_isl.hpp b/src/wrapper/wrap_isl.hpp
index 8b6413e..d930907 100644
--- a/src/wrapper/wrap_isl.hpp
+++ b/src/wrapper/wrap_isl.hpp
@@ -106,6 +106,7 @@ namespace isl
       { \
         if (m_data) \
         { \
+          printf("invalidate %s %d\n", #name, m_data); \
           unref_ctx(get_ctx()); \
           m_data = nullptr; \
         } \
@@ -126,6 +127,7 @@ namespace isl
         if (m_data) \
         { \
           unref_ctx(get_ctx()); \
+          printf("free %s %d\n", #name, m_data); \
           isl_##name##_free(m_data); \
           m_data = nullptr; \
         } \
@@ -137,6 +139,7 @@ namespace isl
         if (data) \
         { \
           m_data = data; \
+          printf("take %s %d\n", #name, m_data); \
           ref_ctx(get_ctx()); \
         } \
       } \
diff --git a/test/test_isl.py b/test/test_isl.py
index d49f6e3..12f3ba2 100644
--- a/test/test_isl.py
+++ b/test/test_isl.py
@@ -414,12 +414,21 @@ def test_remove_map_if_callback():
         m.get_tuple_name(isl.dim_type.in_) == "B")
     assert umap2 == isl.UnionMap.read_from_str(ctx, "{A[0] -> [1]}")
 
+
+def test_remove_map_if_callback_exc():
+    ctx = isl.Context()
+
+    print("CONSTR")
+    umap = isl.UnionMap.read_from_str(ctx, "{A[0] -> [1]; B[1] -> [2]}")
+    print("END CONSTR")
+
     def callback_throws_exception(m):
-        del m
-        assert False
+        1/0
 
     with pytest.raises(isl.Error):
+        print("MAP_IF")
         umap3 = umap.remove_map_if(callback_throws_exception)
+        print("POST MAP_IF")
         del umap3
```
I get this output:
```
CONSTR
take union_map 32282960
END CONSTR
MAP_IF
take union_map 32282960
take map 32738736
[islpy warning] A Python exception occurred in a call back function, ignoring:
Traceback (most recent call last):
  File "test_isl.py", line 426, in callback_throws_exception
    1/0
ZeroDivisionError: division by zero
free map 32738736
free union_map 32282960
free(): double free detected in tcache 2
[1]    218600 abort (core dumped)  PYOPENCL_TEST='port:pthread' python test_isl.py 
```
and valgrind says
```
take union_map 89198416
take union_map 89198416
take map 90189680
[islpy warning] A Python exception occurred in a call back function, ignoring:
Traceback (most recent call last):
  File "test_isl.py", line 426, in callback_throws_exception
    1/0
ZeroDivisionError: division by zero
CONSTR
END CONSTR
MAP_IF
==218966== Invalid read of size 4
==218966==    at 0x65E2CE2: isl_map_free (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x6660CBB: ??? (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65CD443: isl_hash_table_foreach (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x6661534: isl_union_map_free (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x5C512BA: isl::union_map::free_instance() (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5C51259: isl::union_map::~union_map() (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5CFDBAB: std::default_delete<isl::union_map>::operator()(isl::union_map*) const (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5C57DA5: std::unique_ptr<isl::union_map, std::default_delete<isl::union_map> >::~unique_ptr() (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5FD92EA: pybind11::class_<isl::union_map>::dealloc(pybind11::detail::value_and_holder&) (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5A9FDCF: pybind11::detail::clear_instance(_object*) (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5A9FEC9: pybind11_object_dealloc (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x510057: UnknownInlinedFun (object.c:2215)
==218966==    by 0x510057: UnknownInlinedFun (object.h:478)
==218966==    by 0x510057: frame_dealloc.lto_priv.0 (frameobject.c:430)
==218966==  Address 0x5602f70 is 0 bytes inside a block of size 64 free'd
==218966==    at 0x48399AB: free (vg_replace_malloc.c:538)
==218966==    by 0x65E2D5B: isl_map_free (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x5C50F52: isl::map::free_instance() (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5C50EF1: isl::map::~map() (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5CFA5E3: std::default_delete<isl::map>::operator()(isl::map*) const (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5C55CA5: std::unique_ptr<isl::map, std::default_delete<isl::map> >::~unique_ptr() (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5FD8C22: pybind11::class_<isl::map>::dealloc(pybind11::detail::value_and_holder&) (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5A9FDCF: pybind11::detail::clear_instance(_object*) (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5A9FEC9: pybind11_object_dealloc (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x510057: UnknownInlinedFun (object.c:2215)
==218966==    by 0x510057: UnknownInlinedFun (object.h:478)
==218966==    by 0x510057: frame_dealloc.lto_priv.0 (frameobject.c:430)
==218966==    by 0x5262A7: UnknownInlinedFun (object.c:2215)
==218966==    by 0x5262A7: UnknownInlinedFun (object.h:478)
==218966==    by 0x5262A7: UnknownInlinedFun (object.h:541)
==218966==    by 0x5262A7: tb_dealloc.lto_priv.0 (traceback.c:168)
==218966==    by 0x4EE4EE: UnknownInlinedFun (object.c:2215)
==218966==    by 0x4EE4EE: UnknownInlinedFun (object.h:478)
==218966==    by 0x4EE4EE: UnknownInlinedFun (object.h:541)
==218966==    by 0x4EE4EE: insertdict.lto_priv.0 (dictobject.c:1102)
==218966==  Block was alloc'd at
==218966==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==218966==    by 0x65BC945: isl_calloc_or_die (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65E25CA: isl_map_alloc_space (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65E642A: ??? (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65D16DA: ??? (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65D419C: ??? (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65D4720: ??? (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65D5C4E: isl_stream_read_union_map (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x65D602B: isl_union_map_read_from_str (in /usr/lib/x86_64-linux-gnu/libisl.so.22.0.1)
==218966==    by 0x5F68B17: isl::union_map_read_from_str(isl::ctx const&, char*) (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5E8BFED: pybind11::object pybind11::detail::argument_loader<isl::ctx const&, char*>::call_impl<pybind11::object, pybind11::object (*&)(isl::ctx const&, char*), 0ul, 1ul, pybind11::detail::void_type>(pybind11::object (*&)(isl::ctx const&, char*), std::integer_sequence<unsigned long, 0ul, 1ul>, pybind11::detail::void_type&&) && (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
==218966==    by 0x5E432BC: std::enable_if<!std::is_void<pybind11::object>::value, pybind11::object>::type pybind11::detail::argument_loader<isl::ctx const&, char*>::call<pybind11::object, pybind11::detail::void_type, pybind11::object (*&)(isl::ctx const&, char*)>(pybind11::object (*&)(isl::ctx const&, char*)) && (in /home/andreas/src/islpy/islpy/_isl.cpython-38-x86_64-linux-gnu.so)
```
which says to me that isl has freed the `union_map` when there was clearly still a reference being held. (You can see the two references above where `take` occurs twice with the same address.)

I'll disable the test of the error case with a reference to this comment. It would be fantastic if you could report this upstream to Sven.

**Comment by inducer** [comment]
Thanks for working on this!

**Pull request merged**